src/pcm.rs

```diff
@@ -544,6 +568,20 @@ impl<'a> HwParams<'a> {
         acheck!(snd_pcm_hw_params_get_rate(self.0, &mut v, &mut d)).map(|_| v as u32)
     }
 
+    pub fn get_rate_max(&self) -> Result<u32> {
+        let mut v = 0;
+        acheck!(snd_pcm_hw_params_get_rate_max(self.0, &mut v, ptr::null_mut())).map(|_| v as u32)
```

Comment:
The `null_mut` looks weird here. I'm not saying it's wrong, but could you explain why you think this is the right thing to do?

Reply:
Ok after some extensive digging in the alsa source, I can explain this.

The param is actually a return type. If a non-null pointer is supplied, the pointer is dereferenced and the value at that location is updated to inform the caller whether the maximum value is actually included. So basically if there are a range of allowed rates, then after calling snd_pcm_hw_params_get_rate_max the value behind the ptr will be -1 if the set is open (indicating the actual max is less than the returned value), or 0 if closed (indicating that the returned value is included in the interval).

If you pass a null pointer (i.e. 0), then it is ignored: all modifications to the value behind the pointer are preceded with `if (ptr)`.

The documentation is awful for alsa, I will write something in the docs in this PR about all this.

Reply:
I've added some notes about this to the source. They are maths-y but I think that's ok since they won't appear in the docs, they are just there to justify the use of null ptr.